KEYCLOAK-4188 Documentation for Cross DC

matthewhelmke opened the pull request:
This is a large PR with changes across multiple files. I would be most appreciative of a close review.

matthewhelmke committed:
KEYCLOAK-4188 Documentation for Cross DC

matthewhelmke commented:
@hmlnarik and @mposolda , I would appreciate you both checking to make sure I did not change any of the technical details while I was cleaning up the language.

matthewhelmke commented:
@stianst The build failures are coming from URLs used in examples, such as `http://localhost:11080/auth/`. They are necessary for the examples to be useful. Is there a way to mark them up so they are ignored during the build tests?

mposolda commented:
@matthewhelmke Nice! Thanks for the documentation and all the fixes. My vote is to merge the PR.

hmlnarik commented on server_installation/topics/operating-mode/crossdc.adoc:
```diff
@@ -0,0 +1,736 @@
+[[requestprocessing]]
+==== Request processing
+
+An end user's browser sends an HTTP request to the link:{installguide_loadbalancer_link}[front end load balancer]. This load balancer is usually HTTPD or Wildfly with mod_cluster, NGINX, HA Proxy, or perhaps some other kind of software or hardware load balancer.
```
typo: Wild**F**ly

matthewhelmke replied:
Fixed this and a couple more I found.

hmlnarik reviewed (commented):
Thanks! I'll have some fixes to do though, please wait a little before merging.

hmlnarik commented on server_installation/topics/operating-mode/crossdc.adoc:
```diff
@@ -0,0 +1,736 @@
+servers from `site1` are eventually able to read the data saved by {project_name} servers on `site2` . 
+
+Based on the environment, you have the option to decide if you prefer:
+* Reliability - which is typically used in Active/Active mode. Data written on `site1` must be visible immediately on `site2`.
```
Asterisks do not seem to render correctly, please check the build. You might need to insert empty lines between the items (?).

hmlnarik commented on server_installation/topics/operating-mode/crossdc.adoc:
```diff
@@ -0,0 +1,736 @@
+==== Modes
+   
+According your requirements, there are two basic operating modes for Cross-Datacenter Replication:
+* Active/Passive - Here the users and client applications send the requests just to the {project_name} nodes in just a single data center.
```
Asterisks do not seem to render correctly, same as above

matthewhelmke replied:
Nice catch! In all these cases, a blank line is required before the unordered list. I did not know this was an Asciidoc requirement until not. Thank you!

hmlnarik commented on server_installation/topics/operating-mode/crossdc.adoc:
```diff
@@ -0,0 +1,736 @@
+* Active/Active - Here the users and client applications send the requests to the {project_name} nodes in both data centers.
+It means that data need to be visible immediately on both sites and available to be consumed immediately from {project_name} servers on both sites. This is especially true if {project_name} server writes some data on `site1`, and it is required that the data are available immediately for reading by {project_name} servers on `site2` immediately after the write on `site1` is finished.
+
+The active/passive mode is better for performance. For more information about how to configure caches for either mode, see: link:{} [sync or async backups section](#sync-or-async-backups).    
```
the `link:{}` does not render correctly

hmlnarik commented on server_installation/topics/operating-mode/crossdc.adoc:
```diff
@@ -0,0 +1,736 @@
+{project_name} uses a relational database management system (RDBMS) to persist some metadata about realms, clients, users, and so on. See link:{installguide_database_link}[this chapter] of the server installation guide for more details. In a Cross-Datacenter Replication setup, we assume that either both data centers talk to the same database or that every data center has its own database node and ...
+
+Details of DB setup are out-of-scope for {project_name}, however many RDBMS vendors like MariaDB and Oracle offer replicated databases and synchronous replication. We test {project_name} with these vendors:
+* Oracle Database 12c Release 1 (12.1) RAC
```
Asterisks do not seem to render correctly, same as above

hmlnarik commented on server_installation/topics/operating-mode/crossdc.adoc:
```diff
@@ -0,0 +1,736 @@
+.Brute force protection
+
+Finally the `loginFailures` cache is used to track data about failed logins, such as how many times the user `john` 
+entered a bad password. The details are described link:{adminguide_bruteforce_link}[here]. It is up to the admin if he wants this cache to be replicated across data centers. To have an accurate count of login failures, 
```
... if **they** want ... (?)

hmlnarik commented on server_installation/topics/operating-mode/crossdc.adoc:
```diff
@@ -0,0 +1,736 @@
+entered a bad password. The details are described link:{adminguide_bruteforce_link}[here]. It is up to the admin if he wants this cache to be replicated across data centers. To have an accurate count of login failures, 
+the replication is needed. On the other hand, not replicating this data can save some performance. So if performance is more important then accurate counts of login failures, the replication can be avoided. 
+
+For more detail about how caches can be configured see link:tuningcache[Tuning the JDK Cache Configuration].
```
JDK -> JD**G** (too many abbreviations) ;)

hmlnarik commented on server_installation/topics/operating-mode/crossdc.adoc:
```diff
@@ -0,0 +1,736 @@
+[[communication]]
+==== Communication details
+
+{project_name} uses multiple, separate Infinispan clusters. Every {project_name} node is in the cluster 
```
{project_name} uses multiple, separate clusters of Infinispan caches.

hmlnarik commented on server_installation/topics/operating-mode/crossdc.adoc:
```diff
@@ -0,0 +1,736 @@
+==== Communication details
+
+{project_name} uses multiple, separate Infinispan clusters. Every {project_name} node is in the cluster 
+with the other {project_name} nodes in same data center, but not with the {project_name} nodes in different data centers. A {project_name} node does not communicate directly with the {project_name} nodes from different data centers. {project_name} nodes use external JDG (actually Infinispan servers) for communication across data centers. This is done using the link:http://infinispan.org/docs/8.2. ...
```
I'd remove `(actually Infinispan servers)` and choose variant via `{jdgserver_name}`

hmlnarik commented on server_installation/topics/operating-mode/crossdc.adoc:
```diff
@@ -0,0 +1,736 @@
+* The same details apply for `node21` and `node22`. They cluster with each other and communicate only with `jdg2` server using the HotRod protocol.
+  
+Our example setup assumes all 6 servers are bootstrapped on localhost, but each on different ports. It also assumes
+that all 4 {project_name} servers talk to the same database, which can be a locally set MySQL, PostgreSQL, MariaDB 
```
`which can be a locally set MySQL, PostgreSQL, MariaDB or any other` is slightly misleading for it only works for the localhost toy example while we should aim at supported cross-dc dbs which are currently Galera cluster and Oracle RAC. So I'd remove this part completely.

hmlnarik commented on server_installation/topics/operating-mode/crossdc.adoc:
```diff
@@ -0,0 +1,736 @@
+. Download {jdgserver_name} {jdgserver_version} server and unzip to a directory you choose. This location will be referred in later steps as `JDG1_HOME` .
+
+. Change those things in the `JDG1_HOME/standalone/configuration/clustered.xml` in the configuration of JGroups subsystem:
+..Add the `xsite` channel, which will use `tcp` stack, under `channels` element:
```
Does not render correctly

hmlnarik commented on server_installation/topics/operating-mode/crossdc.adoc:
```diff
@@ -0,0 +1,736 @@
+[[setup]]
+==== Basic setup
+
+For this example, we describe using two data centers, `site1` and `site 2`. Each data center consists of 1 Infinispan server and 2 {project_name} servers. We will end up with 2 Infinispan servers and 4 {project_name} servers in total. 
```
extra space in `site 2`
Infinispan -> {jdgserver_name}

On a general note - we say that the example uses two data centers, and then start describing running all of them in localhost. That should be adjusted, either to say that we simulate 2 data centres on localhost or by using different domain. My preference is the latter and I've added all comments necessary to make it possible below, they are marked with `[EX]` prefix.

hmlnarik commented on server_installation/topics/operating-mode/crossdc.adoc:
```diff
@@ -0,0 +1,736 @@
+  
+* The same details apply for `node21` and `node22`. They cluster with each other and communicate only with `jdg2` server using the HotRod protocol.
+  
+Our example setup assumes all 6 servers are bootstrapped on localhost, but each on different ports. It also assumes
```
[EX] `Our example setup assumes all 6 servers are bootstrapped on localhost, but each on different ports.` contradicts the very first sentence in this same section, see above

hmlnarik commented on server_installation/topics/operating-mode/crossdc.adoc:
```diff
@@ -0,0 +1,736 @@
+<stack name="tcp">
+    <transport type="TCP" socket-binding="jgroups-tcp"/>
+    <protocol type="TCPPING">
+        <property name="initial_hosts">localhost[8610],localhost[9610]"</property>
```
[EX] should be `<property name="initial_hosts">jdg1[7600],jdg2[7600]"</property>`

matthewhelmke replied:
Is there an extra " in there after [7600], or should there be another before jdg1?

`<property name="initial_hosts">jdg1[7600],jdg2[7600]"</property>`

hmlnarik commented on server_installation/topics/operating-mode/crossdc.adoc:
````diff
@@ -0,0 +1,736 @@
+</stack>
+```
++
+.. Configure the `tcp` stack to use `TCPPING` protocol instead of `MPING`. Remove the `MPING` element and replace it with the `TCPPING`. The `initial_hosts` element points to the hosts `jdg1` and `jdg2`, which in our example setup are both on `localhost`, but they differ in the port offset:
````
[EX] `, which in our example setup are both on `localhost`, but they differ in the port offset` can be removed.

hmlnarik commented on server_installation/topics/operating-mode/crossdc.adoc:
````diff
@@ -0,0 +1,736 @@
+                
+</cache-container>
+```
+NOTE: Details about the configuration options inside `replicated-cache-configuration` are explained in [later section](#tuning-jdg-cache-configuration). Also
````
link does not render correctly

hmlnarik commented on server_installation/topics/operating-mode/crossdc.adoc:
````diff
@@ -0,0 +1,736 @@
+. In the `JDG2_HOME/standalone/configuration/clustered.xml` exchange `site1` with `site2` and viceversa in the configuration of `relay` in the JGroups subsystem and in configuration of `backups` in the cache-subsystem. For example:
+.. The `relay` element should look like this:
++
+```
````
should be ` ```xml`

hmlnarik commented on server_installation/topics/operating-mode/crossdc.adoc:
````diff
@@ -0,0 +1,736 @@
+
+.. The `backups` element like this:
++
+```
````
should be ` ```xml`

hmlnarik commented on server_installation/topics/operating-mode/crossdc.adoc:
````diff
@@ -0,0 +1,736 @@
+```
+cd JDG1_HOME/bin
+./standalone.sh -c clustered.xml -Djava.net.preferIPv4Stack=true \
+-Djboss.socket.binding.port-offset=1010 -Djboss.default.multicast.address=234.56.78.99 \
````
[EX] the `-Djboss.socket.binding.port-offset=1010` should be removed

hmlnarik commented on server_installation/topics/operating-mode/crossdc.adoc:
````diff
@@ -0,0 +1,736 @@
+```
+cd JDG2_HOME/bin
+./standalone.sh -c clustered.xml -Djava.net.preferIPv4Stack=true \
+-Djboss.socket.binding.port-offset=2010 -Djboss.default.multicast.address=234.56.78.100 \
````
[EX] The -Djboss.socket.binding.port-offset=1010 should be removed

hmlnarik commented on server_installation/topics/operating-mode/crossdc.adoc:
```diff
@@ -0,0 +1,736 @@
+
+. Unzip {project_name} server distribution to a location you choose. It will be referred to later as `NODE11`.
+
+. Configure a shared database for KeycloakDS datasource. It is recommended to use MySQL, MariaDB, or PostgreSQL for testing purposes. See link:database[the Database section] for more details.
```
No Postgres please :)

hmlnarik commented on server_installation/topics/operating-mode/crossdc.adoc:
````diff
@@ -0,0 +1,736 @@
+```
+cd NODE11/bin
+./standalone.sh -c standalone-ha.xml -Djboss.node.name=node11 -Djboss.site.name=site1 \
+-Djboss.default.multicast.address=234.56.78.1 -Dremote.cache.port=12232 -Djava.net.preferIPv4Stack=true \
````
[EX] Replace `-Dremote.cache.port=12232` with `-Dremote.cache.host=jdg1`

hmlnarik commented on server_installation/topics/operating-mode/crossdc.adoc:
```diff
@@ -0,0 +1,736 @@
+cd NODE11/bin
+./standalone.sh -c standalone-ha.xml -Djboss.node.name=node11 -Djboss.site.name=site1 \
+-Djboss.default.multicast.address=234.56.78.1 -Dremote.cache.port=12232 -Djava.net.preferIPv4Stack=true \
+-Djboss.socket.binding.port-offset=3000
```
[EX] remove this line and backslash from the previous line

hmlnarik commented on server_installation/topics/operating-mode/crossdc.adoc:
`````diff
@@ -0,0 +1,736 @@
+````
+cd NODE12/bin
+./standalone.sh -c standalone-ha.xml -Djboss.node.name=node12 -Djboss.site.name=site1 \
+-Djboss.default.multicast.address=234.56.78.1 -Dremote.cache.port=12232 -Djava.net.preferIPv4Stack=true \
`````
[EX] Same as above - Replace `-Dremote.cache.port=12232` with `-Dremote.cache.host=jdg1` and remove the backslash and the following line

hmlnarik commented on server_installation/topics/operating-mode/crossdc.adoc:
````diff
@@ -0,0 +1,736 @@
+```
+cd NODE21/bin
+./standalone.sh -c standalone-ha.xml -Djboss.node.name=node21 -Djboss.site.name=site2 \
+-Djboss.default.multicast.address=234.56.78.2 -Dremote.cache.port=13232 -Djava.net.preferIPv4Stack=true \
````
[EX] Same as above - Replace `-Dremote.cache.port=13232` with `-Dremote.cache.host=jdg2` and remove the backslash and the following line

hmlnarik commented on server_installation/topics/operating-mode/crossdc.adoc:
````diff
@@ -0,0 +1,736 @@
+```
+cd NODE22/bin
+./standalone.sh -c standalone-ha.xml -Djboss.node.name=node22 -Djboss.site.name=site2 \
+-Djboss.default.multicast.address=234.56.78.2 -Dremote.cache.port=13232 -Djava.net.preferIPv4Stack=true \
````
[EX] Same as above - Replace `-Dremote.cache.port=13232` with `-Dremote.cache.host=jdg2` and remove the backslash and the following line

hmlnarik commented on server_installation/topics/operating-mode/crossdc.adoc:
```diff
@@ -0,0 +1,736 @@
+
+. Test:
+
+.. Go to `http://localhost:11080/auth/` and create the initial admin user.
```
[EX] Go to `http://localhost:11080/auth/` -> Go to `http://node11:8080/auth/`

hmlnarik commented on server_installation/topics/operating-mode/crossdc.adoc:
```diff
@@ -0,0 +1,736 @@
+
+.. Go to `http://localhost:11080/auth/` and create the initial admin user.
+
+.. Go to `http://localhost:11080/auth/admin` and login as admin to admin console.
```
[EX] Go to `http://localhost:11080/auth/admin` -> Go to `http://node11:8080/auth/admin`

hmlnarik commented on server_installation/topics/operating-mode/crossdc.adoc:
```diff
@@ -0,0 +1,736 @@
+
+.. Go to `http://localhost:11080/auth/admin` and login as admin to admin console.
+
+.. Open a second browser and go to any of nodes `http://localhost:12080/auth/admin` or `http://localhost:13080/auth/admin` or `http://localhost:14080/auth/admin`. After login, you should be able to see 
```
[EX] `http://localhost:12080/auth/admin` or `http://localhost:13080/auth/admin` or `http://localhost:14080/auth/admin`
->
`http://node12:8080/auth/admin` or `http://node21:8080/auth/admin` or `http://node22:8080/auth/admin`

hmlnarik commented on server_installation/topics/operating-mode/crossdc.adoc:
```diff
@@ -0,0 +1,736 @@
+==== Bringing sites offline and online
+
+For example, assume this scenario:
+. Site `site2` is entirely offline from the `site1` perspective. This means that all JDG servers on `site2` are off *or* the network between `site1` and `site2` is broken. 
```
does not render correctly

hmlnarik commented on server_installation/topics/operating-mode/crossdc.adoc:
```diff
@@ -0,0 +1,736 @@
+. Site `site2` is entirely offline from the `site1` perspective. This means that all JDG servers on `site2` are off *or* the network between `site1` and `site2` is broken. 
+. You run {project_name} servers and JDG server `jdg1` in site `site1` 
+. Someone logs in on a {project_name} server on `site1`. 
+. The {project_name} server from `site1` will try to write the session to the remote cache on `jdg1` server, which is supposed to backup data to the `jdg2` server in the `site2`. See [this section](#communication-details) for the details. 
```
link does not render

hmlnarik commented on server_installation/topics/operating-mode/crossdc.adoc:
```diff
@@ -0,0 +1,736 @@
+. The {project_name} server from `site1` will try to write the session to the remote cache on `jdg1` server, which is supposed to backup data to the `jdg2` server in the `site2`. See [this section](#communication-details) for the details. 
+. Server `jdg2` is offline or unreachable from `jdg1`. So the backup from `jdg1` to `jdg2` will fail.
+. The exception is thrown in `jdg1` log and the failure will be propagated from `jdg1` server to {project_name} servers as well because 
+the default `FAIL` backup failure policy is configured. See [this section](#backup-failure-policy) for details around the backup policies.
```
link does not render

hmlnarik commented on server_installation/topics/operating-mode/crossdc.adoc:
```diff
@@ -0,0 +1,736 @@
+This is useful especially if the outage is planned. With `jconsole` or CLI, you can connect to the `jdg1` server and take the `site2` offline. 
+More details about this are available in the link:https://access.redhat.com/documentation/en-us/red_hat_jboss_data_grid/7.1/html/administration_and_configuration_guide/set_up_cross_datacenter_replication#taking_a_site_offline[JDG documentation]. 
+ 
+WARNING: This has turned off the backup to `site2` for the cache `sessions`. The same steps usually need to be done for all the other {project_name} caches mentioned [here](#sync-or-async-backups).
```
link does not render

hmlnarik commented on server_installation/topics/operating-mode/crossdc.adoc:
```diff
@@ -0,0 +1,736 @@
+Again, you may need to check all the caches and bring them online.
+
+Once the sites are put online, it's usually good to:
+* Do the link:statetransfer[state transfer].
```
items do not render

hmlnarik commented on server_installation/topics/operating-mode/crossdc.adoc:
```diff
@@ -0,0 +1,736 @@
+[[statetransfer]]
+==== State transfer
+
+State transfer is a manually required step. JDG does not do this automatically, for example during split-brain, it is only the admin who may decide which site has preference and hence if state transfer needs to be done bidirectionally between both sites or just unidirectionally, as in only from `site1` to `site2`, but not from `site2` to `site1`. 
```
"manually required step" - should it be "required manual step"?
JDG -> {jdgserver_name}

hmlnarik commented on server_installation/topics/operating-mode/crossdc.adoc:
```diff
@@ -0,0 +1,736 @@
+
+A bidirectional state transfer will ensure that entities which were created *after* split-brain on `site1` will be transferred to `site2`. This is not an issue as they do not yet exist on `site2`. Similarly, entities created *after* split-brain on `site2` will be transferred to `site1`. Possibly problematic parts are those entities which exist *before* split-brain on both sites and which were upd ...
+
+Unfortunately, there is no any universal solution to this. Split-brains and network outages are just state, which is usually impossible to be handled 100% correctly with 100% consistent data between sites. In the case of {project_name}, it typically is not a critical issue. In the worst case, users will need to re-login again to their clients, or have the improper count of loginFailures tracked f ...
```
docmentation -> documentation

matthewhelmke replied:
Hahahahaha! How embarrassing for the tech writer to misspell "documentation". LOL

hmlnarik commented on server_installation/topics/operating-mode/crossdc.adoc:
```diff
@@ -0,0 +1,736 @@
+
+Unfortunately, there is no any universal solution to this. Split-brains and network outages are just state, which is usually impossible to be handled 100% correctly with 100% consistent data between sites. In the case of {project_name}, it typically is not a critical issue. In the worst case, users will need to re-login again to their clients, or have the improper count of loginFailures tracked f ...
+
+The state transfer can be also done on the JDG side through JMX. The operation name is `pushState`. There are few other operations to monitor status, cancel push state, and so on. 
```
JDG -> {jdgserver_name}

hmlnarik commented on server_installation/topics/operating-mode/crossdc.adoc:
```diff
@@ -0,0 +1,736 @@
+Unfortunately, there is no any universal solution to this. Split-brains and network outages are just state, which is usually impossible to be handled 100% correctly with 100% consistent data between sites. In the case of {project_name}, it typically is not a critical issue. In the worst case, users will need to re-login again to their clients, or have the improper count of loginFailures tracked f ...
+
+The state transfer can be also done on the JDG side through JMX. The operation name is `pushState`. There are few other operations to monitor status, cancel push state, and so on. 
+More info about state transfer is available in the link:https://access.redhat.com/documentation/en-us/red_hat_jboss_data_grid/7.1/html/administration_and_configuration_guide/set_up_cross_datacenter_replication#state_transfer_between_sites[JDG docs].
```
JDG -> {jdgserver_name}

hmlnarik commented on server_installation/topics/operating-mode/crossdc.adoc:
```diff
@@ -0,0 +1,736 @@
+
+To clear the caches, see {adminguide_clearcache_link}[{adminguide_clearcache_name}].
+
+When the network is back, it is sufficient to clear the cache just on one {project_name} node on any random site. The cache invalidation event will be sent to all the other {project_name} nodes in all sites. However, it needs to be done for all the caches (realms, users, keys).    
```
Reference to where to find in in admin console would help (hint: server_admin/topics/realms/cache.adoc file)

hmlnarik commented on server_installation/topics/operating-mode/crossdc.adoc:
```diff
@@ -0,0 +1,736 @@
+
+The difference between `FAIL` and `WARN` is that when `FAIL` is used and the JDG server tries to back data up to the other site and the backup fails then the failure will be propagated back to the caller (the {project_name} server). The backup might fail because the second site is temporarily unreachable or there is a concurrent transaction which is trying to update same entity. In this case, the ...
+
+When using `WARN`, the failed backups are not propagated from the JDG server to the {project_name} server. The user won't see the error and the failed backup will be just ignored. There will be a shorter timeout, typically 10 seconds as that's the default timeout for backup. It can be changed by the attribute `timeout` of `backup` element. There won't be retries. There will just be a WARNING mess ...
```
JDG -> {jdgserver_name}

hmlnarik commented on server_installation/topics/operating-mode/crossdc.adoc:
```diff
@@ -0,0 +1,736 @@
+ 
+With timeout 0, the transaction will immediately fail and then will be retried from {project_name} if backup `failure-policy` with the value `FAIL` is configured. As long as the second concurrent transaction is finished, the retry will usually be successful and the entity will have applied updates from both concurrent transactions.
+
+We see very good consistency and results for concurrent transaction with this configuration, at it is recommended to keep it. 
```
at it -> and it

hmlnarik commented on server_installation/topics/operating-mode/crossdc.adoc:
```diff
@@ -0,0 +1,736 @@
+
+We see very good consistency and results for concurrent transaction with this configuration, at it is recommended to keep it. 
+
+The only (non-functional) problem is the exception in the JDG log, which happens every time when the lock is not
```
JDG -> {jdgserver_name}

hmlnarik commented on server_installation/topics/operating-mode/crossdc.adoc:
```diff
@@ -0,0 +1,736 @@
+==== SYNC or ASYNC backups
+
+An important part of the `backup` element is the `strategy` attribute. You must decide whether it needs to be `SYNC` or `ASYNC`. We have 7 caches which might be Cross-Datacenter Replication aware, and these can be configured in 3 different modes regarding cross-dc:
+. SYNC backup
```
does not render correctly

hmlnarik commented on server_installation/topics/operating-mode/crossdc.adoc:
````diff
@@ -0,0 +1,736 @@
+Generally, if you are in doubt and performance is not a blocker for you, it's safer to keep the caches in `SYNC` strategy.
+
+WARNING: Regarding the switch to SYNC/ASYNC backup, make sure that you edit the `strategy` attribute of the the `backup` element. For example like this:
+```
````
\``` -> ` ```xml`

hmlnarik commented on server_installation/topics/operating-mode/crossdc.adoc:
````diff
@@ -0,0 +1,736 @@
+```
+<backup site="site2" failure-policy="FAIL" strategy="ASYNC" enabled="true">
+```
+Not the `mode` attribute of cache-configuration element.
````
Not the `mode` attribute of cache-configuration element.
->
Note the `mode` attribute of `cache-configuration` element.

hmlnarik commented on server_installation/topics/operating-mode/crossdc.adoc:
```diff
@@ -0,0 +1,736 @@
+
+* It is recommended to go through the link:setup[example setup] and have this one working first, so that you have some understanding of how things work. It is also wise to read this entire document to have some understanding of things.
+
+* Check in jconsole cluster status (GMS) and the JGroups status (RELAY) of JDG as described in link:jdgsetup[JDG server setup]. If things no not look as expected, then the issue is likely in the setup of JDG servers. 
```
no not look -> do not look
JDG -> {jdgserver_name}

hmlnarik commented on server_installation/topics/operating-mode/crossdc.adoc:
````diff
@@ -0,0 +1,736 @@
+
+```
++
+it usually means that {project_name} server is not able to reach the JDG server in his own datacenter. Make sure that firewall is set as expected and JDG server is possible to connect. 
````
JDG -> {jdgserver_name}

hmlnarik commented on server_installation/topics/operating-mode/crossdc.adoc:
````diff
@@ -0,0 +1,736 @@
+ ...
+```
++
+then check the log of corresponding JDG server of your site and check if has failed to backup to the other site. If the backup site is unavailable, then it is recommended to switch it offline, so that JDG server won't try to backup to the offline site causing the operations to pass successfully on {project_name} server side as well. See link:administration[Administration of Cross-DC Deployment] f ...
````
JDG -> {jdgserver_name}

hmlnarik reviewed (changes requested):
First of all, thanks a lot for the docs.
I've updated the images in the gdrive to match the datacenter names from the text, please update them.
Please see inline for further comments.

hmlnarik commented:
@mposolda Could you please check whether the [EX] match your expectations?

hmlnarik commented:
@matthewhelmke Ad links, these should be ignored if you add them to [src/keycloak-documentation/tests/src/test/resources/ignored-links](https://github.com/keycloak/keycloak-documentation/blob/master/tests/src/test/resources/ignored-links)

hmlnarik commented:
The examples are not yet tested but I'll test them ASAP. This should not block addressing the [EX] comments, I'll send a fix commit if I find anything wrong with them.

mposolda commented:
+1 to Hynek's comment. We agreed that it will be better to address the EX comments and use the hosts instead of localhost+port-offsets .

hmlnarik commented:
> Is there an extra " in there after [7600], or should there be another
before jdg1?
Good catch, there should be no quote inside the <property> tag

<property name="initial_hosts">jdg1[7600],jdg2[7600]</property>

On Wed, Nov 29, 2017 at 1:49 PM, Matthew Helmke <notifications@github.com>
wrote:

> *@matthewhelmke* commented on this pull request.
> ------------------------------
>
> In server_installation/topics/operating-mode/crossdc.adoc
> <https://github.com/keycloak/keycloak-documentation/pull/248#discussion_r153777355>
> :
>
> > +<stack name="udp">
> +    ...
> +    <relay site="site1">
> +        <remote-site name="site2" channel="xsite"/>
> +        <property name="relay_multicasts">false</property>
> +    </relay>
> +</stack>
> +```
> ++
> +.. Configure the `tcp` stack to use `TCPPING` protocol instead of `MPING`. Remove the `MPING` element and replace it with the `TCPPING`. The `initial_hosts` element points to the hosts `jdg1` and `jdg2`, which in our example setup are both on `localhost`, but they differ in the port offset:
> ++
> +```xml
> +<stack name="tcp">
> +    <transport type="TCP" socket-binding="jgroups-tcp"/>
> +    <protocol type="TCPPING">
> +        <property name="initial_hosts">localhost[8610],localhost[9610]"</property>
>
> Is there an extra " in there after [7600], or should there be another
> before jdg1?
>
> <property name="initial_hosts">jdg1[7600],jdg2[7600]"</property>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/keycloak/keycloak-documentation/pull/248#discussion_r153777355>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFJEAL2ALhl3VX59snYH549DNzMzFEMSks5s7VLGgaJpZM4QtLLJ>
> .
>



-- 

--Hynek

matthewhelmke commented:
@hmlnarik I've incorporated all your changes (great stuff in your review, BTW).

I've also added the example URLs into the ignored-links file--thank you for helping me learn about that!

hmlnarik commented:
@matthewhelmke Thank you! Few comments:
- Could you please update images so that use "site1" and "site2" labels (they are updated in gdrive, see above)
- The ignored-links file seem to be accidentally omitted from the code
- The "Variables not found" issue from the bot needs to be addressed (maybe some other brackets would fix it)
- When these are addressed, please squash the commits into a single one

matthewhelmke commented:
@hmlnarik Thanks again for catching my errors (most of which I should have caught myself...)!

I've included the fixes in the most recent commit.

How do I squash the commits into a single one?

matthewhelmke commented:
Hmm...I have also created a conflict. :-/
